File: java/479_Largest _Palindrome_Product.java
/*
https://leetcode.com/problems/largest-palindrome-product/
Find the largest palindrome made from the product of two n-digit numbers.
Since the result could be very large, you should return the largest palindrome mod 1337.

For a given n find the largest possible number which is a product. Take the first half of the digits of this
number and reverse them to create the largest possible palindrome. Test if this palindrome is a product by checking
factors with n digits starting from 10**n - 1 and stopping at sqrt(palindrome). If not, decrement the first half.
Time - O(10**2n), 10**n firstHalf to check, each of which is checked for 10**n factors
Space - O(n)
*/

public class Solution {
    public int largestPalindrome(int n) {

        if(n == 1)
            return 9;

        int upperBound = (int) Math.pow(10, n) - 1;     // factors must be <= upperBound
        int lowerBound = (upperBound / 10);             // factors must be > lowerBound
        long maxNumber = (long) upperBound * (long) upperBound;     // max possible palindrome

        // first half of the maximum possible palindrome
        int firstHalf = (int) (maxNumber / (long) Math.pow(10, n));

        boolean palindromeFound = false;
        long palindrome = 0;

        while (!palindromeFound) {
            // create maximum possible palindrome
            palindrome = createPalindrome(firstHalf);

            // try all possible factors i and palindrome/i
            for (long i = upperBound; upperBound > lowerBound; i--) {
                // stop if largest factor squared cannot make palindrome
                if (i * i < palindrome)
                    break;

                // two factors found, both are n-digits
                if (palindrome % i == 0) {
                    palindromeFound = true;
                    break;
                }
            }

            firstHalf--;
        }
        return (int) (palindrome % 1337);
    }

    private long createPalindrome(long num) {
        String str = num + new StringBuilder().append(num).reverse().toString();
        return Long.parseLong(str);
    }
}

File: python_1_to_1000/479_Largest_Palindrome_Product.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/largest-palindrome-product/
# Find the largest palindrome made from the product of two n-digit numbers.
# Since the result could be very large, you should return the largest palindrome mod 1337.

# Build palindromes with 2 * n digits, starting from the largest palindrome and check if they can be factorised.
# Define palindrome as upper * 10**n + lower.
# The two factors are M = 10**n - i and L = 10**n - j where i and j are small.
# Hence palindrome = (10**n - i) * (10**n - j) = 10**n * (10**n - (i + j)) + i * j = upper * 10**n + lower.
# So upper = 10**n - (i + j) and lower = i * j.
# Define a = i + j. This means that a = 10**n - upper.
# Substituting for j = a - i, the equation for lower can be rewritten as lower = a * i - i * i.
# Solving for i, in order for i to be an integer sqrt(a**2 - 4 * lower) must be an integer.
# Time - O(n * 10**n) for the for loop
# Space - O(1)

class Solution(object):
    def largestPalindrome(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n == 1:
            return 9

        for a in range(1, 9 * 10 ** (n - 1)):

            hi = (10 ** n) - a          # most significant digits of palindrome
            lo = int(str(hi)[::-1])     # least significant digits of palindrome

            if a ** 2 - 4 * lo < 0:
                continue

            if (a ** 2 - 4 * lo) ** .5 == int((a ** 2 - 4 * lo) ** .5):     # sqrt(a**2 - 4*lo) is an integer
                return (lo + 10 ** n * (10 ** n - a)) % 1337
